src/qttools/datastructures/dsbcoo.py

```diff
@@ -52,6 +52,79 @@ def __init__(
         self.rows = xp.asarray(rows).astype(int)
         self.cols = xp.asarray(cols).astype(int)
 
+    def _normalize_index(self, index: tuple) -> tuple:
+        """Adjusts the sign to allow negative indices and checks bounds."""
+        if not isinstance(index, tuple):
+            raise IndexError("Invalid index.")
+
+        if not len(index) == 2:
+            raise IndexError("Invalid index.")
+
+        row, col = index
+
+        row = self.shape[-2] + row if row < 0 else row
+        col = self.shape[-1] + col if col < 0 else col
+        if not (0 <= row < self.shape[-2] and 0 <= col < self.shape[-1]):
+            raise IndexError("Index out of bounds.")
+
+        return row, col
+
+    def __getitem__(self, index: tuple) -> ArrayLike:
+        """Gets a single value or from the data structure."""
+        row, col = self._normalize_index(index)
+        ind = xp.where((self.rows == row) & (self.cols == col))[0]
+
+        if self.distribution_state == "stack":
+            if len(ind) == 0:
+                return xp.zeros(self.data.shape[:-1], dtype=self.dtype)
+
+            return self.data[..., ind[0]]
+
+        # If nnz are distributed accross the ranks, we need to find the
+        # rank that holds the data.
+        if len(ind) == 0:
+            # We cannot know which rank is supposed to hold an element
+            # that is not in the matrix, so we raise an error.
+            raise IndexError("Requested element not in matrix.")
```

Comment:
We can maybe discuss at some point if this is the correct thing to do in this specific case (`nnz` distribution and accessing a zero element)